File: JavaScript/AOC2018/days/day03.js
import fs from "fs"
import path from "path"

export default class Day3 {
    constructor() {
        const filePath = path.resolve("../../Data/2018/day3.txt");
        const regex = /[0-9]+/g;
        this.data = fs.readFileSync(filePath, "utf-8")
            .split(/\r?\n/)
            .map(line => line.trim())
            .filter(line => line !== "")
            .map(line => {
                const matches = line.match(regex);
                return matches ? matches.map(Number) : [];
        });
        this.grid = new Int32Array(1000000);
        for (const arr of this.data) {
            this.markFabric(arr);
        }
    }

    solve() {
        console.log("Day 3 solutions:");
        this.partOne();
        this.partTwo();
    }

    partOne() {
        let count = 0;
        for (let i = 0; i < this.grid.length; ++i) {
            if (this.grid[i] === -1)
                count++;
        }
        console.log(`There number of overlaps is: ${count}`);
    }

    partTwo() {
        for (const arr of this.data) {
            if (this.checkOverlap(arr)) {
                console.log(`The fabric with no overlap has id ${arr[0]}`);
                return;
            }
        }
    }

    markFabric(arr) {
        for (let i = arr[1]; i < arr[1] + arr[3]; ++i) {
            for (let j = arr[2]; j < arr[2] + arr[4]; ++j) {
                const idx = 1000 * i + j;
                if (this.grid[idx] === 0) {
                    this.grid[idx] = arr[0];
                } else {
                    this.grid[idx] = -1;
                }
            }
        }
    }

    checkOverlap(arr) {
        for (let i = arr[1]; i < arr[1] + arr[3]; ++i) {
            for (let j = arr[2]; j < arr[2] + arr[4]; ++j) {
                const idx = 1000 * i + j;
                if (this.grid[idx] === -1) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: Python/AOC2018/days/day03.py
import numpy as np
import re

class Day3:
    def __init__(self):
        with open("../../Data/2018/day3.txt") as f:
            self.data = [[int(n) for n in re.findall(r"\d+", line)] for line in f if line.strip()]
        self.grid = np.zeros((1000, 1000), dtype=np.int32)

    def solve(self):
        for claim in self.data:
            self.mark_fabric(claim)
        print("Day 3 solutions:")
        self.part_one()
        self.part_two()

    def part_one(self):
        print(f"The total overlaps are: {np.count_nonzero(self.grid == -1)}")

    def part_two(self):
        for claim_id, x, y, w, h in self.data:
            if np.all(self.grid[y:y+h, x:x+w] == claim_id):
                print(f"The non-overlapping claim ID is: {claim_id}")

    def mark_fabric(self, claim):
        claim_id, x, y, w, h = claim
        y_end, x_end = y + h, x + w
        region = self.grid[y:y_end, x:x_end]
        self.grid[y:y_end, x:x_end] = np.where(region == 0, claim_id, -1)